### PR TITLE
Drop QDQ around more nodes

### PR DESCRIPTION
**Pull request by mcollinswisc**
### Description

Extends the Drop QDQ optimization to remove DequantizeLinear and QuantizeLinear nodes from around operators:

- Flatten
- Expand
- Tile
- Slice
- GatherElements
- ReduceMin
- ReduceMax

### Motivation and Context

To reduce floating-point conversions in quantize inference. Mainly motivated by the Flatten case, since that will show up in graphs exported from PyTorch to ONNX. But to make the change complete, extending to a larger set of ops for which this optimization is valid.

https://github.com/microsoft/onnxruntime/issues/21375

### CODE REVIEW

**Comment by mcollinswisc** [comment]
Changes are on top of https://github.com/microsoft/onnxruntime/pull/21182, since it also needs to check for positive scale when dropping QDQ around ReduceMin and ReduceMax

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc [hunk 1]
```diff
@@ -46,26 +48,49 @@ void DropQDQNodesRules(SelectorActionRegistry& qdq_selector_action_registry) {
 
   std::unique_ptr<Action> drop_action_no_int16 = std::make_unique<MergeIntoTargetFixed>(
       std::vector<NodeAndMoveInfo>(moves));  // Copy before std::move(moves)
+  std::unique_ptr<Action> drop_action_no_nonpositive_scale = std::make_unique<MergeIntoTargetFixed>(
+      std::vector<NodeAndMoveInfo>(moves));  // Copy before std::move(moves)
+  std::unique_ptr<Action> drop_action_no_int16_nor_nonpositive_scale = std::make_unique<MergeIntoTargetFixed>(
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc:53]: drop_action_no_int16_and_positive_scale?

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc [hunk 2]
```diff
@@ -46,26 +48,49 @@ void DropQDQNodesRules(SelectorActionRegistry& qdq_selector_action_registry) {
 
   std::unique_ptr<Action> drop_action_no_int16 = std::make_unique<MergeIntoTargetFixed>(
       std::vector<NodeAndMoveInfo>(moves));  // Copy before std::move(moves)
+  std::unique_ptr<Action> drop_action_no_nonpositive_scale = std::make_unique<MergeIntoTargetFixed>(
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc:51]: Can we avoid the double negative any maybe have `drop_action_has_positive_scale`?

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc [hunk 3]
```diff
@@ -46,26 +48,49 @@ void DropQDQNodesRules(SelectorActionRegistry& qdq_selector_action_registry) {
...
+  // of the elements between quantized & de-quantized values.
   std::unique_ptr<NodeSelector> selector_disallow_16bit = std::make_unique<QDQ::DropQDQNodesSelector>(false);
   qdq_selector_action_registry.RegisterSelectorAndAction(drop_action_no_int16_name,
-                                                         {{"MaxPool", {12}},
-                                                          {"Resize", {}}},
+                                                         {{"Resize", {}}},
                                                          std::move(selector_disallow_16bit),
                                                          std::move(drop_action_no_int16));
 
+  std::unique_ptr<NodeSelector> selector_disallow_16bit_and_nonpositive_scale = std::make_unique<QDQ::DropQDQNodesSelector>(false, true, false);
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc:72]: nit: please keep lines under 120 chars

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc [hunk 4]
```diff
@@ -46,26 +48,49 @@ void DropQDQNodesRules(SelectorActionRegistry& qdq_selector_action_registry) {
...
+                                                          {"GatherElements", {}},
                                                           {"Reshape", {}},
                                                           {"Transpose", {}},
                                                           {"Squeeze", {}},
-                                                          {"Unsqueeze", {}}},
+                                                          {"Unsqueeze", {}},
+                                                          {"Flatten", {}},
+                                                          {"Expand", {}},
+                                                          {"Tile", {}},
+                                                          {"Slice", {}}},
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc:93]: nit: alphabetically sorted would be slightly nicer

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc (hunk 1 above)
Comment (mcollinswisc) [reply onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc]: Change made in base PR https://github.com/microsoft/onnxruntime/pull/21182

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc (hunk 2 above)
Comment (mcollinswisc) [reply onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc]: I think this action ended up being removed

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc (hunk 4 above)
Comment (mcollinswisc) [reply onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc]: Done in https://github.com/microsoft/onnxruntime/pull/21376/commits/848f701408614a6ef99be2d0bbe856bfd671f9be

& thanks for taking a look!

**Comment by mcollinswisc** [comment]
I had missed the reason why the windows builds were failing last few commits (currently don't have a Windows system to try locally)

I guess since it's

```
onnxruntime\test\optimizer\qdq_transformer_test.cc(1,1): Error C1128: number of sections exceeded object file format limit: compile with /bigobj
```

I'll try adding the `/bigobj` flag similarly to here:
https://github.com/microsoft/onnxruntime/blob/b9f3a5d5b62d12fb3d90b3e98bf7d05aa8a560d6/cmake/onnxruntime_unittests.cmake#L892

**Review by skottmckay — APPROVED** [review]

**Pull request merged**